attach.go

```diff
@@ -0,0 +1,155 @@
+package mail
+
+import (
+	"encoding/base64"
+	"errors"
+	"io/ioutil"
+	"mime"
+	"path/filepath"
+)
+
+// File represents the file that can be added to the email message.
+// You can add attachment from file in path, from base64 string or from []byte.
+// You can define if attachment is inline or not.
+type File struct {
+	// FilePath is the path of the file to attach.
+	FilePath string
+	// Name is the name of file in attachment. Required for Data and B64File. Optional for FilePath.
+	Name string
+	// MimeType of attachment. If empty then is obtained from Name (if not empty) or FilePath. If cannot obtained, application/octet-stream is set.
+	MimeType string
+	// B64File is the base64 string of file to attach.
+	B64File string
+	// Data is the []byte of file to attach.
+	Data []byte
+	// Inline defines if attachment is inline or not.
+	Inline bool
+}
+
+type attachType int
+
+const (
+	attachData attachType = iota
+	attachB64
+	attachFile
+)
+
+// Attach allows you to add an attachment to the email message.
+// The attachment can be inlined
+func (email *Email) Attach(file *File) *Email {
+	if email.Error != nil {
+		return email
+	}
+
+	var name = file.Name
+	var mimeType = file.MimeType
+
+	// if no alternative name was provided, get the filename
+	if len(name) == 0 && len(file.FilePath) > 0 {
+		_, name = filepath.Split(file.FilePath)
+	}
+
+	// get the mimetype
+	if mimeType == "" {
+		mimeType = mime.TypeByExtension(filepath.Ext(name))
+		if mimeType == "" {
+			mimeType = "application/octet-stream"
+		}
+	}
+
+	attachTy, err := getAttachmentType(file)
+	if err != nil {
+		email.Error = errors.New("Mail Error: Failed to add attachment with following error: " + err.Error())
+		return email
+	}
+
+	file.Name = name
+	file.MimeType = mimeType
+
+	switch attachTy {
+	case attachData:
+		email.attachData(file)
+	case attachB64:
+		email.Error = email.attachB64(file)
+	case attachFile:
+		email.Error = email.attachFile(file)
+	}
+
+	return email
+}
+
+func getAttachmentType(file *File) (attachType, error) {
+	// 1- data
+	// 2- base64
+	// 3- file
+
+	// first check if Data
+	if len(file.Data) > 0 {
+		// data requires a name
+		if len(file.Name) == 0 {
+			return 0, errors.New("Mail Error: Attach from bytes requires a name")
+		}
+		return attachData, nil
```

Comment:
Is it ok to have multiple data fields populated or should it be an error condition if more than one of `FilePath`, `B64File` and `Data` contain data?

Reply:
Yes, it's ok to have multiple, but only will be use one and is selected in this order:

- `Data`
- `B64File`
- `FilePath`

So, if `FilePath` and `B64File` is set, this last is used and `FilePath` is ignored.

Maybe a doc is need it in the struct.

Reply:
Yes. It is a good idea to document the precedence order.

I am a bit worried about the possibility that the implementation will silently skip erroneous input. This could mask a bug in the application using this library.

Reply:
This implementation is used a lot in aws-sdk-go. For example, to index an image to rekognition, you can use bytes or S3 Object image https://github.com/aws/aws-sdk-go/blob/v1.38.27/service/rekognition/api.go#L11972-L11982

I don't know if return an error when use multiple or use an order, but I don't think the package can skip erroneous input because the `getAttachmentType` function can handle this. Any "silent error" is a bad implementation of developer.

I will document the precedence order. And left this conversation open some days. If we find a bug I will update. Any check let me know.

Reply:
I decide close this conversation because is not a breaking change and developers, at any problem, can use the old way instead.